### PR TITLE
added HarmonizationNotebook required functions

### PR DESCRIPTION
**Pull request by chenlingantelope**
I added a harmonization directory in scvi/ to include 2 python scripts 

1. utils.py 
- data loading functions for the harmonization notebook dataset
- a wrapper for louvain clustering
- functions for plotting gene expression on umap

2. DEfunctions.py
- compute differentially expressed genes comparing one cluster to all other clusters
- compute differentially expressed genes comparing the cells from 2 conditions within the same cluster

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,3 +1,7 @@
+umap==0.3.7
+seaborn==0.9.0
```
Comment (romain-lopez) [inline requirements.txt]: Can we delete all these requirements ?

File: scvi/harmonization/DEfunctions.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
+import numpy as np
+import pandas as pd
+
+use_cuda = True
+
+
+def GetScale(trainer, n_batches, n_samples_per_batch, gene_dataset, selection, seed=0):
+    scale = []
+    np.random.seed(seed)
+    for batch in range(n_batches):
```
Comment (romain-lopez) [inline scvi/harmonization/DEfunctions.py]: Some of these functions should be encapsultated in the posterior object themselves and not as a new module

File: scvi/harmonization/DEfunctions.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        posterior = trainer.create_posterior(trainer.model, gene_dataset, indices=idx)
+        scale.append(posterior.sequential().get_harmonized_scale(batch))
+    scale = np.concatenate(scale)
+    return scale
+
+
+def GetBayesFactor(scale1, scale2, m_permutation):
+    res = np.repeat(0, scale1.shape[1])
+    for _ in range(m_permutation):
+        sample1 = scale1[np.random.choice(np.arange(scale1.shape[0]), 1), :].ravel()
```
Comment (romain-lopez) [inline scvi/harmonization/DEfunctions.py]: That looks like replicate code, which we do not want

File: scvi/harmonization/DEfunctions.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    res = np.log(res + 1e-8) - np.log(1 - res + 1e-8)
+    return res
+
+
+def ComputeDE(trainer, gene_dataset, idx1, idx2, batch1, batch2, nrep=1, n_samples_per_batch=1000,
+              m_permutation=10000, seed=0):
+    bayes = []
+    np.random.seed(seed)
+    seeds = np.random.choice(1000000, nrep)
+    for i in np.arange(0, nrep):
```
Comment (romain-lopez) [inline scvi/harmonization/DEfunctions.py]: same here

File: scvi/harmonization/utils.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+import pandas as pd
+from scvi.dataset.dataset import GeneExpressionDataset
+import igraph as ig
+import louvain
+from sklearn.neighbors import kneighbors_graph
+
+
+def get_matrix_from_dir(dirname, storage='../'):
+    geneid = loadtxt(storage + dirname + '/genes.tsv', dtype='str', delimiter="\t")
+    cellid = loadtxt(storage + dirname + '/barcodes.tsv', dtype='str', delimiter="\t")
```
Comment (romain-lopez) [inline scvi/harmonization/utils.py]: that should already be in the package somewhere ?

File: scvi/harmonization/utils.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    count = mmread(storage + dirname + '/matrix.mtx')
+    return count, geneid, cellid
+
+
+class ExpressionMatrix(GeneExpressionDataset):
+    def __init__(self, filename, save_path):
+        self.save_path = save_path
+        self.filename = filename
+        count, barcode, genenames = self.preprocess()
+        super(ExpressionMatrix, self).__init__(
```
Comment (romain-lopez) [inline scvi/harmonization/utils.py]: Why do we need these objects ?

File: scvi/harmonization/utils.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        return count, barcode, genenames
+
+
+def louvain_clusters(latent, k=10, rands=0):
+    nn_matrix = kneighbors_graph(latent, k)
+    rows, cols = np.where(nn_matrix.todense() == 1)
+    edges = [(row, col) for row, col in zip(rows, cols)]
+    g = ig.Graph()
+    g.add_vertices(latent.shape[0])
+    g.add_edges(edges)
```
Comment (romain-lopez) [inline scvi/harmonization/utils.py]: Can we use scanpy instead for example ? Or do we really need to cluster ?

File: scvi/harmonization/utils.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+def plot_marker_genes(latent_u, count, genenames, markers):
+    nrow = (len(markers) // 3 + 1)
+    figh = nrow * 4
+    plt.figure(figsize=(10, figh))
+    for i, x in enumerate(markers):
+        exprs = count[:, genenames == x].ravel()
+        idx = (exprs > 0)
+        plt.subplot(nrow, 3, (i + 1))
+        plt.scatter(latent_u[:, 0], latent_u[:, 1], c='lightgrey', edgecolors='none', s=5)
```
Comment (romain-lopez) [inline scvi/harmonization/utils.py]: Should we refactor this code as part of the graph utils.py ?

File: scvi/harmonization/utils.py (hunk 6 above)
Comment (jeff-regier) [reply scvi/harmonization/utils.py]: Should this be in the `dataset` directory? (Or can we delete it entirely?)

**Pull request closed without merging**